### PR TITLE
feat(recordings): include jvmId in Notifications to -web

### PR DESCRIPTION
**Pull request by mwangggg**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] [Signed all commits using a GPG signature](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#gpg-commit-signature-verification)

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes: #1452
Related to https://github.com/cryostatio/cryostat-web/pull/1126
Depends on #1743 (for ease of review/verifying event message contents)

### CODE REVIEW

**Review by tthvo — COMMENTED** [review]
Looks good! Tho, I am not sure if this solves #1452. There are other resources to look into:
```ts
'activeRecordings',
'archivedRecordings',
'eventTemplates',
'eventTypes',
'agentProbes',
```

References: https://github.com/cryostatio/cryostat-web/blob/86a8094c1947004fb1f625da7f517d071e2920d1/src/app/Topology/Entity/types.ts#L43

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6578832827

**Comment by andrewazores** [comment]
Rebase please

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6579065965

**Comment by andrewazores** [comment]
> Looks like it makes sense, though I wonder if it's worth doing some refactoring at this point to extract a reusable utility method that looks something like `notificationFactory.createOwnedResourceBuilder(targetId)`, which still returns a notification builder but pre-configured with the JSON content type, a message Map with the target and jvmId pre-populedand can have the owned resource appended, etc.

Here's a rough and incomplete patch illustrating a bit more of this half-baked idea I presented before:

```patch
diff --git a/src/main/java/io/cryostat/messaging/notifications/Notification.java b/src/main/java/io/cryostat/messaging/notifications/Notification.java
index 44a4f6ea..6fbce23d 100644
--- a/src/main/java/io/cryostat/messaging/notifications/Notification.java
+++ b/src/main/java/io/cryostat/messaging/notifications/Notification.java
@@ -15,6 +15,9 @@
  */
 package io.cryostat.messaging.notifications;
 
+import java.util.HashMap;
+import java.util.Map;
+
 import io.cryostat.net.web.http.HttpMimeType;
 
 public class Notification<T> {
@@ -43,7 +46,7 @@ public class Notification<T> {
     }
 
     public static class Builder<T> {
-        private final NotificationSource source;
+        protected final NotificationSource source;
         private String category = "generic";
         private MetaType type = new MetaType(HttpMimeType.JSON);
         private T message;
@@ -82,6 +85,27 @@ public class Notification<T> {
         }
     }
 
+    public static class OwnedResourceBuilder extends Builder<Map<String, Object>> {
+
+        private final Map<String, Object> map = new HashMap<>();
+
+        OwnedResourceBuilder(NotificationSource source, String category) {
+            super(source);
+            metaType(HttpMimeType.JSON);
+            metaCategory(category);
+            message(map);
+        }
+
+        public OwnedResourceBuilder messageEntry(String key, Object value) {
+            if (value == null) {
+                this.map.remove(key);
+            } else {
+                this.map.put(key, value);
+            }
+            return this;
+        }
+    }
+
     public static class Meta {
         private final String category;
         private final MetaType type;
diff --git a/src/main/java/io/cryostat/messaging/notifications/NotificationFactory.java b/src/main/java/io/cryostat/messaging/notifications/NotificationFactory.java
index 1167e17f..c929db2e 100644
--- a/src/main/java/io/cryostat/messaging/notifications/NotificationFactory.java
+++ b/src/main/java/io/cryostat/messaging/notifications/NotificationFactory.java
@@ -15,23 +15,35 @@
  */
 package io.cryostat.messaging.notifications;
 
-import io.cryostat.net.web.http.HttpMimeType;
+import io.cryostat.recordings.JvmIdHelper;
+import io.cryostat.recordings.JvmIdHelper.JvmIdGetException;
 
 public class NotificationFactory {
 
     private final NotificationSource source;
+    private final JvmIdHelper jvmIdHelper;
 
-    NotificationFactory(NotificationSource source) {
+    NotificationFactory(NotificationSource source, JvmIdHelper jvmIdHelper) {
         this.source = source;
+        this.jvmIdHelper = jvmIdHelper;
     }
 
     public <T> Notification.Builder<T> createBuilder() {
         return new Notification.Builder<T>(source);
     }
 
-    public <T> Notification.Builder<T> createOwnedResourceBuilder(String notificationCategory) {
-        return new Notification.Builder<T>(source)
-                .metaType(HttpMimeType.JSON)
-                .metaCategory(notificationCategory);
+    public Notification.OwnedResourceBuilder createOwnedResourceBuilder(
+            String notificationCategory) {
+        return new Notification.OwnedResourceBuilder(source, notificationCategory);
+    }
+
+    public Notification.OwnedResourceBuilder createOwnedResourceBuilder(
+            String targetId, String notificationCategory) throws JvmIdGetException {
+        return new Notification.OwnedResourceBuilder(source, notificationCategory)
+                // FIXME the websocket notification system should only emit the targetId under one
+                // named key or the other, not both - are we already consistent with this?
+                .messageEntry("target", targetId)
+                .messageEntry("connectUrl", targetId)
+                .messageEntry("jvmId", jvmIdHelper.getJvmId(targetId));
     }
 }
diff --git a/src/main/java/io/cryostat/messaging/notifications/NotificationsModule.java b/src/main/java/io/cryostat/messaging/notifications/NotificationsModule.java
index 9777bfff..120f72c3 100644
--- a/src/main/java/io/cryostat/messaging/notifications/NotificationsModule.java
+++ b/src/main/java/io/cryostat/messaging/notifications/NotificationsModule.java
@@ -19,6 +19,8 @@ import java.util.Set;
 
 import javax.inject.Singleton;
 
+import io.cryostat.recordings.JvmIdHelper;
+
 import dagger.Lazy;
 import dagger.Module;
 import dagger.Provides;
@@ -34,7 +36,8 @@ public abstract class NotificationsModule {
 
     @Provides
     @Singleton
-    static NotificationFactory provideNotificationFactory(NotificationSource source) {
-        return new NotificationFactory(source);
+    static NotificationFactory provideNotificationFactory(
+            NotificationSource source, JvmIdHelper idHelper) {
+        return new NotificationFactory(source, idHelper);
     }
 }
diff --git a/src/main/java/io/cryostat/net/web/http/api/beta/RecordingsFromIdPostHandler.java b/src/main/java/io/cryostat/net/web/http/api/beta/RecordingsFromIdPostHandler.java
index 6bd995d3..87dd745a 100644
--- a/src/main/java/io/cryostat/net/web/http/api/beta/RecordingsFromIdPostHandler.java
+++ b/src/main/java/io/cryostat/net/web/http/api/beta/RecordingsFromIdPostHandler.java
@@ -47,6 +47,7 @@ import io.cryostat.net.web.http.api.ApiVersion;
 import io.cryostat.net.web.http.api.v2.ApiException;
 import io.cryostat.recordings.JvmIdHelper;
 import io.cryostat.recordings.JvmIdHelper.JvmIdDoesNotExistException;
+import io.cryostat.recordings.JvmIdHelper.JvmIdGetException;
 import io.cryostat.recordings.RecordingArchiveHelper;
 import io.cryostat.recordings.RecordingMetadataManager;
 import io.cryostat.recordings.RecordingMetadataManager.Metadata;
@@ -274,35 +275,30 @@ public class RecordingsFromIdPostHandler extends AbstractAuthenticatedRequestHan
                                     try {
 
                                         notificationFactory
-                                                .createOwnedResourceBuilder(NOTIFICATION_CATEGORY)
-                                                .message(
-                                                        Map.of(
-                                                                "recording",
-                                                                new ArchivedRecordingInfo(
-                                                                        connectUrl,
-                                                                        fsName,
-                                                                        webServer
-                                                                                .get()
-                                                                                .getArchivedDownloadURL(
-                                                                                        connectUrl,
-                                                                                        fsName),
-                                                                        webServer
-                                                                                .get()
-                                                                                .getArchivedReportURL(
-                                                                                        connectUrl,
-                                                                                        fsName),
-                                                                        metadata,
-                                                                        size,
-                                                                        archivedTime),
-                                                                "target",
+                                                .createOwnedResourceBuilder(
+                                                        connectUrl, NOTIFICATION_CATEGORY)
+                                                .messageEntry(
+                                                        "recording",
+                                                        new ArchivedRecordingInfo(
                                                                 connectUrl,
-                                                                "jvmId",
-                                                                jvmId))
+                                                                fsName,
+                                                                webServer
+                                                                        .get()
+                                                                        .getArchivedDownloadURL(
+                                                                                connectUrl, fsName),
+                                                                webServer
+                                                                        .get()
+                                                                        .getArchivedReportURL(
+                                                                                connectUrl, fsName),
+                                                                metadata,
+                                                                size,
+                                                                archivedTime))
                                                 .build()
                                                 .send();
                                     } catch (URISyntaxException
                                             | UnknownHostException
-                                            | SocketException e) {
+                                            | SocketException
+                                            | JvmIdGetException e) {
                                         logger.error(e);
                                         throw new ApiException(500, e);
                                     }
```

WDYT? This is just meant to make it a little easier to pass the right data in and get the expected notification emitted, or rather to make it a little harder to forget to include things.

**Comment by mwangggg** [comment]
looks good! I'll try it out

**Comment by andrewazores** [comment]
> looks good! I'll try it out

https://github.com/mwangggg/cryostat/pull/20

**Comment by mwangggg** [comment]
Is it better to include  jvmIdHelper in notificationFactory and notificationModule, or to pass in jvmId to createOwnedResourceBuilder since some of the methods where createOwnedResourceBuilder already initialized jvmId

**Comment by mwangggg** [comment]
there are some notifications that emit `targetId` and not `jvmId`- none that are involved in this PR- but I can just call `createOwnedResourceBuilder(notificationCategory)` then manually add all the message entries if we want them to call a new method instead of `createBuilder`. I'm unsure if there's any notifications that emit mismatched `targetId` and `jvmId` though. It makes sense if the goal is to minimize user error.

**Comment by andrewazores** [comment]
> there are some notifications that emit targetId and not jvmId

In those cases I think the rough draft I submitted is still fine. Whatever component is listening for those notifications should not be affected if the server starts to include the `jvmId` on those messages, so long as the message isn't changing in some other structural way.

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6589917416

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6590033116

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6590060020

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6590078303

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6591016522

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6613848721

File: src/main/java/io/cryostat/messaging/notifications/Notification.java [hunk 1]
```diff
@@ -84,10 +85,30 @@ public Notification<T> build() {
...
+                this.map.put(key, value);
+            }
+            return this;
+        }
+    }
+
     public static class Meta {
         private final String category;
         private final MetaType type;
-        private final long serverTime = Instant.now().getEpochSecond();
```
[the remark this answers is not in the document]
Comment (mwangggg) [reply src/main/java/io/cryostat/messaging/notifications/Notification.java:90]: https://github.com/cryostatio/cryostat-web/pull/1126/commits/b684474df052d66d53321377fa98a223b82ee520

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6630848597

File: src/main/java/io/cryostat/net/reports/AbstractReportGeneratorService.java [hunk 2]
```diff
@@ -85,8 +85,7 @@ Path copyRecordingToFile(
                 continue;
             }
             try (OutputStream out = new BufferedOutputStream(new FileOutputStream(path.toFile()))) {
-                try (conn;
-                        InputStream in = conn.getService().openStream(rec, false)) {
+                try (InputStream in = conn.getService().openStream(rec, false)) {
```
Comment (andrewazores) [inline src/main/java/io/cryostat/net/reports/AbstractReportGeneratorService.java:88]: What was the reason for this change? It doesn't seem related to the rest of the patch.

Comment (mwangggg) [reply src/main/java/io/cryostat/net/reports/AbstractReportGeneratorService.java:88]: it was just because spotbug was failing ...

Comment (andrewazores) [reply src/main/java/io/cryostat/net/reports/AbstractReportGeneratorService.java:88]: Weird - it doesn't seem to normally fail here. I wonder what made it pick this up as a bug. Was it complaining about the `conn` being a null pointer? I've seen that one before, it's some false positive that happens when you use `try-with-resources` blocks.

Speaking of which though, this nesting looks silly anyway:

```java
            try (OutputStream out = new BufferedOutputStream(new FileOutputStream(path.toFile()))) {
                try (conn;
                        InputStream in = conn.getService().openStream(rec, false)) {
                    byte[] buff = new byte[READ_BUFFER_SIZE];
                    int n = 0;
                    while ((n = in.read(buff)) != -1) {
                        out.write(buff, 0, n);
                        if (!targetConnectionManager.markConnectionInUse(cd)) {
                            throw new IOException(
                                    "Target connection unexpectedly closed while streaming"
                                            + " recording");
                        }
                    }
                    out.flush();
                    return path;
                }
            }
```

Looks like those two `try` blocks can be combined into one.

**Comment by andrewazores** [comment]
I just noticed while reviewing this that I don't see the websocket messages being written in the server logs anymore, but that is also true on `main`. They are supposed to get logged here: https://github.com/cryostatio/cryostat/blob/fd2b69a3f4f893a79f3626cee845ae5c9d81b9a5/src/main/java/io/cryostat/messaging/MessagingServer.java#L214

Do you see log messages containing the websocket events?

File: src/main/java/io/cryostat/net/reports/AbstractReportGeneratorService.java (hunk 2 above)
Comment (mwangggg) [reply src/main/java/io/cryostat/net/reports/AbstractReportGeneratorService.java:88]: yea that's what failed it. I'll add it back in and merge the try blocks

**Comment by andrewazores** [comment]
Great:

```
c1da58d1f1c28b39b2cad0d2fc71741759297db7 is the first bad commit
commit c1da58d1f1c28b39b2cad0d2fc71741759297db7
Author: Andrew Azores <aazores@redhat.com>
Date:   Thu Oct 12 15:52:44 2023 -0400

    fix(build): use vertx-dependencies and netty-bom for version management (#1720)

 pom.xml | 32 ++++++++++++++++++++++----------
 1 file changed, 22 insertions(+), 10 deletions(-)
```

#1720 somehow breaks some of our logging.

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6631741944

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat/actions/runs/6631783636

**Comment by andrewazores** [comment]
Rebase please

**Review by andrewazores — APPROVED** [review]
